File: backend/src/main/java/com/fptgang/backend/model/File.java
package com.fptgang.backend.model;


import com.fptgang.backend.util.Searchable;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.apache.commons.lang3.RandomStringUtils;
import org.apache.commons.lang3.RandomUtils;
import org.hibernate.annotations.CreationTimestamp;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "files")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class File {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long fileId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "uploader_id", nullable = false)
    private Account uploader;

    @Column(columnDefinition = "NVARCHAR(255)", length = 255, nullable = false)
    @Searchable
    private String fileName;

    @Column(columnDefinition = "NVARCHAR(255)", length = 255, nullable = false)
    @Searchable
    private String fileUrl;

    @Column(nullable = false)
    private String fileType;

    @Column(nullable = false)
    private Long size;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @Column(nullable = false, columnDefinition = "BOOLEAN DEFAULT TRUE")
    @Builder.Default
    private Boolean isVisible = true;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "proposal_id")
    private Proposal proposal;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_id")
    private Project project;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "message_id")
    private Message message;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "milestone_id")
    private Milestone milestone;

    @NotNull
    public static File.FileBuilder of(MultipartFile blob) {
        return File.builder()
            .fileName(blob.getOriginalFilename())
            .fileType(blob.getContentType())
            .size(blob.getSize())
            .isVisible(true);
    }
}

File: gendb/src/engine/model/File.ts
import {escapeSingleQuotes} from "../utils.js";

export class File {
  file_id: number;
  created_at: Date | null;
  file_name: string;
  file_type: string;
  file_url: string;
  is_visible: boolean;
  size: number;
  message_id: number | null;
  project_id: number | null;
  proposal_id: number | null;
  uploader_id: number;
  milestone_id: number | null;

  constructor(data: Partial<File>) {
    this.file_id = data.file_id || 0;
    this.created_at = data.created_at || null;
    this.file_name = data.file_name || '';
    this.file_type = data.file_type || '';
    this.file_url = data.file_url || '';
    this.is_visible = data.is_visible !== undefined ? data.is_visible : true;
    this.size = data.size || 0;
    this.message_id = data.message_id || null;
    this.project_id = data.project_id || null;
    this.proposal_id = data.proposal_id || null;
    this.uploader_id = data.uploader_id || 0;
    this.milestone_id = data.milestone_id || null;
  }

  static dump(files: File[]): string {
    if (!files.length) return '';

    const fields = [
      'file_id',
      'created_at',
      'file_name',
      'file_type',
      'file_url',
      'is_visible',
      'size',
      'message_id',
      'project_id',
      'proposal_id',
      'uploader_id',
      'milestone_id'
    ];

    const values = files.map(file => {
      return `(${[
        file.file_id,
        file.created_at ? `'${file.created_at.toISOString().slice(0, 19).replace('T', ' ')}'` : 'NULL',
        `'${escapeSingleQuotes(file.file_name)}'`,
        `'${escapeSingleQuotes(file.file_type)}'`,
        `'${escapeSingleQuotes(file.file_url)}'`,
        file.is_visible ? 1 : 0,
        file.size,
        file.message_id || 'NULL',
        file.project_id || 'NULL',
        file.proposal_id || 'NULL',
        file.uploader_id,
        file.milestone_id || 'NULL'
      ].join(', ')})`;
    });

    return `INSERT INTO \`files\` (${fields.map(f => '`' + f + '`').join(', ')})
            VALUES ${values.join(',\n')};`;
  }
}